core/src/main/java/org/apache/accumulo/core/client/impl/TabletServerBatchWriter.java

```diff
@@ -763,7 +763,7 @@ public void run() {
               log.trace("{} - binning {} mutations", Thread.currentThread().getName(),
                   mutationsToSend.size());
               addMutations(mutationsToSend);
-            } catch (Exception e) {
+            } catch (Throwable e) {
               updateUnknownErrors("Error processing mutation set", e);
```

Comment:
This change very dangerously allows the JVM to continue working as though nothing serious happened when an Error occurred, and that is not a good idea. This pattern of behavior was addressed in newer versions in #1840 . While 1.10 still has that behavior throughout, I think this moves it further in the direction away from where we are headed with 2.1. I'm not sure what a better solution would be, but I wouldn't want the JVM to continue processing mutations as though nothing was wrong, if there were an Error.

Reply:
The Throwable is passed through to the client via updateUnknownErrors where it is re-thrown. The client can decide to halt the JVM if they so choose.

Reply:
@andrewglowacki wrote:
> The Throwable is passed through to the client via updateUnknownErrors where it is re-thrown. The client can decide to halt the JVM if they so choose.

From what I see, only the lastUnknownError is preserved. This might just get logged and never sent back to the calling client code. The Error can easily get suppressed by a more recent and less serious exception, and the serious Error go overlooked indefinitely.

@keith-turner wrote:
> it seems reasonable that the library should make a best effort attempt to get any errors that occur in that thread back to the user code

I probably agree that is a reasonable goal, but I'm not convinced the proposed change does that. I only had time for a quick code review, and it looked like this change was more likely to result in the Errors being *less* noticeable to users than *more* noticeable. However, it's not completely suppressed, as I had originally thought. There is a chance the Error will make it back to the user code via a MutationsRejectedException. I don't think it's quite as "failsafe" as killing the JVM or triggering a user-defined UncaughtExceptionHandler, but it might be the best option for 1.10.

I'd feel more comfortable about this if I knew that the Error couldn't be completely clobbered by a less severe Exception and that it definitely would make it back to the user code. Perhaps something can be done with [`addSuppressed`](https://docs.oracle.com/javase/7/docs/api/java/lang/Throwable.html#addSuppressed(java.lang.Throwable)) ?

> It seems like the argument against this is that it may not work therefore do not even make the attempt.

I don't recognize my position in your portrayal here.

Reply:
>> It seems like the argument against this is that it may not work therefore do not even make the attempt.
>
> I don't recognize my position in your portrayal here.

That was a synthesis of how I felt after reading lots of comments around this.  However the statement is not grounded in any fact and throwing it in was not productive. Please forgive the mistake and ignore the comment.

Reply:
@ctubbsii I see what you mean. I agree it is possible the OOM error is never seen if for example a previously executing SendTask encounters a different exception or error. I believe this would be the case if the error occurred in SendTask though as well. 

I think using addSupressed here makes sense, or a queue of Throwables waiting to be thrown to he user via addSupressed. I can however also imagine a case where that is never seen either. For example, if SendTask evokes a normal server exception and then OOM occurs in a different SendTask after the user retrieves the original exception.

Reply:
<details>
<summary>@keith-turner No worries.</summary>
Reading back through #2331, I can see how my strongly-worded opinions there would prime someone into thinking I'd be against attempting to solve this here. While I may have some strong philosophical opinions on how to handle these kinds of Errors in general, I just want to be clear that I'm not opposed to attempting a practical workaround here.
</details>

@andrewglowacki I think you're right that there are still other cases where they wouldn't be returned. Problems queued by threads after the user got a previous problem is one case (though these might be seen later, during a `bw.flush()` or `bw.close()`). Another case might be if a `RuntimeException` or `Error` were to be thrown from the outer thread (main thread?) that fell through to the calling code and prevented the queued problems from being sent back.

Additionally, merely adding suppressed mutations might cause the same problems to reported multiple times, unless we clear them after they get sent to the calling code. For example, if `bw.addMutation(m)` were wrapped with try/catch block instead of a single catch clause at the end of a try-with-resources block for the `BatchWriter`. However, clearing problems isn't great because, currently, the failure state is never reset for the BW when problem occurred. So, it appears that the current code could throw the last error multiple times. For example, `addMutation` could throw it, and then, since it's not reset, it could throw again at `flush` and/or `close`. If we clear the queue of problems to avoid the same problems reported multiple times, that might cause the second throw of `MutationsRejectedException` to not have an actual problem to wrap (because it might have been cleared, even though `somethingFailed` is still `true`). I'm not really sure the preferred behavior here. It's more complicated than it seems at first glance, because we're queuing up these problems instead of handling them immediately. It's uncertain how best to handle it.